app/gen-server/lib/HomeDBManager.ts

```diff
@@ -2479,6 +2490,85 @@ export class HomeDBManager extends EventEmitter {
       .getOne() || undefined;
   }
 
+  // can be factorized with getShareByLinkId
+  public async getDocApiKeyByLinkId(docId: string, linkId: string): Promise<Share | undefined> {
+    return await this._connection.createQueryBuilder()
+      .select('shares')
+      .from(Share, 'shares')
+      .where('shares.doc_id = :docId', {docId})
+        .andWhere('shares.link_id = :linkId', {linkId})
+      .getOne() || undefined;
+  }
+
+  public async getDocApiKeyByKey(key: string): Promise<Share | null> {
+    return await this.getShareByKey(key);
+  }
+
+  public async createDocApiKey(docId: string, share: ShareInfo) {
+    const key = makeId();
+    const query = await this._connection.createQueryBuilder()
+      .insert()
+      .setParameter('options', share.options)
+      .into(Share)
+      .values({
+        linkId: share.linkId,
+        docId: docId,
+        options: JSON.parse(share.options),
+        key,
+      })
+      .execute() || undefined;
+    return query ? key : query;
+  }
+
+  // in parameters linkId is the linkId in db in case of update of this id in the share
+  public async updateDocApiKeyByLinkId(docId: string, linkId: string, share: ShareInfo) {
+    return await this._connection.createQueryBuilder()
+      .update(Share)
+      .set(share)
+      .where('doc_id = :docId and link_id = :linkId', {docId, linkId})
+      .execute() || undefined;
+  }
+
+  public async updateDocApiKeyByKey(docId: string, apiKey: string, share: ShareInfo) {
+    return await this._connection.createQueryBuilder()
+      .update(Share)
+      .set(share)
+      .where('doc_id = :docId and key = :apiKey', {docId, apiKey})
+      .execute() || undefined;
+  }
+
+  public async deleteDocApiKeyByKey(docId: string, apiKey: string) {
+    return await this.connection.createQueryBuilder()
+      .delete()
+      .from('shares')
+      .where('doc_id = :docId and key = :apiKey', {docId, apiKey})
+      .execute() || undefined;
+  }
+
+  public async getDocApiKeys(docId: string): Promise<Share[] | undefined> {
+    return await this._connection.createQueryBuilder()
+      .select('shares')
+      .from(Share, 'shares')
+      .where('doc_id = :docId', {docId})
+      .getMany() || undefined;
+  }
+
+  public async deleteDocApiKeyByLinkId(docId: string, linkId: string) {
+    return await this.connection.createQueryBuilder()
+      .delete()
+      .from('shares')
+      .where('doc_id = :docId and link_id = :linkId', {docId, linkId})
+      .execute() || undefined;
+  }
+
+  public async deleteDocApiKeys(docId: string) {
+    return await this.connection.createQueryBuilder()
+      .delete()
+      .from('shares')
+      .where('doc_id = :docId', {docId})
+      .execute() || undefined;
+  }
```

Comment:
Maybe you may move this whole part (and everything that CRUDs Shares) in its own module, just like the work around [the Users](https://github.com/gristlabs/grist-core/blob/57cacc9e2f081f5e7128bc51c800ecf21e324c08/app/gen-server/lib/homedb/UsersManager.ts)?

NB: however `getOrgOwnerFromDocApiKey` should remain in HomeDBManager, it actually retrieves an org.

Reply:
Done